### PR TITLE
Add Kdoc to pagination related classes, and improve usage doc

### PR DESCRIPTION
**Pull request by BoD**

### CODE REVIEW

File: libraries/apollo-normalized-cache-api-incubating/src/commonMain/kotlin/com/apollographql/apollo3/cache/normalized/api/MetadataGenerator.kt [hunk 1]
```diff
@@ -3,39 +3,72 @@ package com.apollographql.apollo3.cache.normalized.api
...
 
+/**
+ * Used to associate arbitrary metadata to objects.
+ * For example, information about pagination can later be used to merge pages (see [RecordMerger]).
+ *
+ * The metadata is stored and attached to the object's field in the [Record] resulting from the normalization.
+ * For instance, given the query `query MyQuery { foo }` and an implementation of [metadataForObject] returning
+ * `mapOf("key", 0)`, the resulting Record will look like `fields: { foo: bar }, metadata: { foo: { key: 0 } }`.
+ *
+ * See also [Record.metadata].
```
Comment (martinbonnin) [inline libraries/apollo-normalized-cache-api-incubating/src/commonMain/kotlin/com/apollographql/apollo3/cache/normalized/api/MetadataGenerator.kt:16]:
```suggestion
 * @see [Record.metadata]
```

File: libraries/apollo-normalized-cache-api-incubating/src/commonMain/kotlin/com/apollographql/apollo3/cache/normalized/api/MetadataGenerator.kt [hunk 2]
```diff
@@ -3,39 +3,72 @@ package com.apollographql.apollo3.cache.normalized.api
 import com.apollographql.apollo3.annotations.ApolloExperimental
 import com.apollographql.apollo3.api.CompiledField
 import com.apollographql.apollo3.api.Executable
+import com.apollographql.apollo3.api.json.ApolloJsonElement
 
+/**
+ * Used to associate arbitrary metadata to objects.
+ * For example, information about pagination can later be used to merge pages (see [RecordMerger]).
```
Comment (martinbonnin) [inline libraries/apollo-normalized-cache-api-incubating/src/commonMain/kotlin/com/apollographql/apollo3/cache/normalized/api/MetadataGenerator.kt:10]:
```suggestion
 * A generator for arbitrary metadata associated with objects.
 * For example, information about pagination can later be used to merge pages (see [RecordMerger]).
```

File: libraries/apollo-normalized-cache-api-incubating/src/commonMain/kotlin/com/apollographql/apollo3/cache/normalized/api/RecordMerger.kt [hunk 3]
```diff
@@ -1,12 +1,26 @@
 package com.apollographql.apollo3.cache.normalized.api
 
 import com.apollographql.apollo3.annotations.ApolloExperimental
+import com.apollographql.apollo3.api.json.ApolloJsonElement
+import com.apollographql.apollo3.cache.normalized.api.FieldRecordMerger.FieldMerger
 
+/**
+ * Used to merge incoming [Record]s from the network with existing ones in the cache.
```
Comment (martinbonnin) [inline libraries/apollo-normalized-cache-api-incubating/src/commonMain/kotlin/com/apollographql/apollo3/cache/normalized/api/RecordMerger.kt:8]:
```suggestion
 * A merger that merges incoming [Record]s from the network with existing ones in the cache.
```

File: libraries/apollo-normalized-cache-api-incubating/src/commonMain/kotlin/com/apollographql/apollo3/cache/normalized/api/RecordMerger.kt [hunk 4]
```diff
@@ -1,12 +1,26 @@
...
+   * Merges the incoming Record with the existing Record.
+   *
+   * @param newDate optional date to associate with the fields of the resulting merged Record. If null, a date will not be set.
+   * @return a pair of the resulting merged Record and a set of field keys which have changed or were added.
+   */
   fun merge(existing: Record, incoming: Record, newDate: Long?): Pair<Record, Set<String>>
 }
 
+/**
+ * Default [RecordMerger] that merges fields by replacing them with the incoming fields.
```
Comment (martinbonnin) [inline libraries/apollo-normalized-cache-api-incubating/src/commonMain/kotlin/com/apollographql/apollo3/cache/normalized/api/RecordMerger.kt:22]:
```suggestion
 * A [RecordMerger] that merges fields by replacing them with the incoming fields.
```

File: libraries/apollo-normalized-cache-api-incubating/src/commonMain/kotlin/com/apollographql/apollo3/cache/normalized/api/RecordMerger.kt [hunk 5]
```diff
@@ -37,17 +51,35 @@ object DefaultRecordMerger : RecordMerger {
   }
 }
 
+/**
+ * Convenience implementation of [RecordMerger] that simplifies the merging of [Record]s by delegating to a [FieldMerger].
```
Comment (martinbonnin) [inline libraries/apollo-normalized-cache-api-incubating/src/commonMain/kotlin/com/apollographql/apollo3/cache/normalized/api/RecordMerger.kt:55]:
```suggestion
 * Convenience implementation of [RecordMerger] that simplifies the merging of [Record]s by delegating to a [FieldMerger].
```
```suggestion
 * A convenience implementation of [RecordMerger] that simplifies the merging of [Record]s by delegating to a [FieldMerger].
```

**Review by martinbonnin — APPROVED** [review]
Very nice additions, I added a bunch of bikesheds but LGTM!

It could be interesting to refine the terminology around Record, cachekey, field name, field key, arguments, etc... and update the rest of the code appropriately.

File: design-docs/Normalized cache pagination.md [hunk 6]
```diff
@@ -228,7 +230,7 @@ With that in place, after fetching the first page, the cache will look like this
 | user:1     | id: 1, name: John Smith                                 |
 | user:2     | id: 2, name: Jane Doe                                   |
 
-Notice how the cache key no longer includes the `page` argument, which means watching `UsersPage(page = 1)` (or any page) is enough to observe the whole list.
+The field name no longer includes the `page` argument, which means watching `UsersPage(page = 1)` or any page will observe the same list.
```
[the remark this answers is not in the document]
Comment (BoD) [reply design-docs/Normalized cache pagination.md:233]: Good point! I like "field key". This means `FieldNameGenerator` should be renamed too.

[the remark this answers is not in the document]
Comment (martinbonnin) [reply design-docs/Normalized cache pagination.md:234]: Yea, which is breaking...

**Pull request merged**